eeg_flow/tasks/ica_decomposition.py

```diff
@@ -0,0 +1,658 @@
+        del locks
+
+
+def apply_ica(
+    participant: str,
+    group: str,
+    task: str,
+    run: int,
+    *,
+    timeout: float = 10,
+    overwrite: bool = False,
+):
+    """Apply the reviewed ICA decomposition.
+
+    Parameters
+    ----------
+    %(participant)s
+    %(group)s
+    %(task)s
+    %(run)s
+    %(timeout)s
+    overwrite : bool
+        If True, overwrites existing derivatives.
+    """
+    check_type(overwrite, (bool,), "overwrite")
+    # prepare folders
+    _, derivatives_folder_root, username = load_config()
+    derivatives_folder = get_derivative_folder(
+        derivatives_folder_root, participant, group, task, run
+    )
+    fname_stem = get_fname(participant, group, task, run)
+
+    # lock the output derivative files
+    derivatives = (derivatives_folder / f"{fname_stem}_step6_preprocessed_raw.fif",)
+    locks = lock_files(*derivatives, timeout=timeout)
+    try:
+        # The raw saved after interpolation of bridges already contains bad channels and
+        # segments. No need to reload the "info" and "oddball_with_bads" annotations.
+        raw = read_raw_fif(
+            derivatives_folder / f"{fname_stem}_step2_raw.fif", preload=True
+        )
+
+        # apply ICA for mastoids
+        raw_mastoids = raw.copy()
+        raw_mastoids.filter(
+            l_freq=0.5,
+            h_freq=40.0,
+            picks="eeg",
+            method="fir",
+            phase="zero-double",
+            fir_window="hamming",
+            fir_design="firwin",
+            pad="edge",
+        )
+        ica = read_ica(derivatives_folder / f"{fname_stem}_step5_reviewed_1st_ica.fif")
+        ica.apply(raw_mastoids)
+        del ica  # free resources
+        raw_mastoids.pick(["M1", "M2"])
+
+        # trick MNE in thinking that a custom-ref has been applied
+        with raw_mastoids.info._unlock():
+            raw_mastoids.info["custom_ref_applied"] = FIFF.FIFFV_MNE_CUSTOM_REF_ON
+
+        # apply ICA for EEG channels
+        raw.drop_channels(["M1", "M2"])
+        raw.filter(
+            l_freq=0.5,
+            h_freq=40.0,
+            picks="eeg",
+            method="fir",
+            phase="zero-double",
+            fir_window="hamming",
+            fir_design="firwin",
+            pad="edge",
+        )
+        raw.set_montage(None)  # just in case we have a montage left
+        raw.add_reference_channels(ref_channels="CPz")
+        raw.set_eeg_reference("average", projection=False)
+        ica = read_ica(derivatives_folder / f"{fname_stem}_step5_reviewed_2nd_ica.fif")
+        ica.apply(raw)
+        del ica  # free resources
+
+        raw.set_eeg_reference(["CPz"], projection=False)  # change reference back
+        raw.add_channels([raw_mastoids])
+        del raw_mastoids
+        raw.set_montage("standard_1020")  # add montage for non-mastoids
+        raw.set_eeg_reference(["M1", "M2"])
+
+        # plot for inspection
+        raw.plot(theme="light", block=True)
```

Comment:
@ArthurNguyen I suggest that during this last plot you also check the "bad segment" annotated, and edit them based on what was well corrected and what was not. This way, in the last `_step6_preprocessed_raw.fif` file, you  get a clean raw recording with correct annotations.

Reply:
yes, makes a lot of sense to me!